### PR TITLE
integration: refactor regression test framework

### PR DESCRIPTION
**Pull request by JFixby**
### All Submissions:

* [ ] Submission follows the [Code Contribution Guidelines](../blob/master/docs/code_contribution_guidelines.md)
* [x] There are not any other open [Pull Requests](../pulls) for the same update/change
* [ ] Commit messages are formatted according to [Model Git Commit Messages](../blob/master/docs/code_contribution_guidelines.md#44-model-git-commit-messages)
* [x] All changes are compliant with the latest version of Go and the one prior to it
* [ ] The code being submitted is commented according to the [Code Documentation and Commenting](../blob/master/docs/code_contribution_guidelines.md#CodeDocumentation) section of the Code Contribution Guidelines
* [ ] Any new logging statements use an appropriate subsystem and logging level
* [x] Code has been formatted with go fmt
* [x] Running go test does not fail any tests or report any vet issues
* [ ] Running golint does not report any new issues that did not already exist

### New Feature Submissions:

* [x] Code is accompanied by tests which exercise both the positive and negative (error paths) conditions (if applicable)

### Bug Fixes:

* [x] Code is accompanied by new tests which trigger the bug being fixed to prevent regressions

### Changes to Core Features:

* [x] An explanation of what the changes do and why they should be included is provided
* [x] Code is accompanied by updates to tests and/or new tests for the core changes (if applicable)

The purpose of the PR is to implement regression testing framework for `dcrd` and `dcrwallet`.

### Proposed changes:
 - Refactored package: `integration/`
Provides registry to ensure proper tracking and disposal of leaky assets.
Provides helpers for the test setup to detect its own defects.
Offers temporary directories management.
Offers management for reusable resources (e.g harnesses).
Test files were moved to `integration/harness/simpleregtest/`

 - Edit: package `main`
`rpcserver_test.go` was moved to `integration/harness/simpleregtest/`

 - Added package: `integration/commandline/`
Provides helpers to run external command-line tools, e.g `go` code builder, `dcrd` and `dcrwallet` executables.
Ensures proper disposal of the external processes to avoid CPU leaks.
See `example_test.go` for the usage.

 - Added package: `integration/gobuilder/`
Helps to build a target Go project upon request.

- `integration/rpctest/`-package was completely demolished.
Tests were split into separated files and moved to `integration/harness/simpleregtest/`
Harnesses and test setup code was refactored and moved to `integration/harness`

- Added package: `integration/harness/`
Provides reusable modules allowing to assemble and to configure any desired test environment.
Defines test harness and it's sub-components.

- Added package: `integration/harness/memwallet/`
Offers a simple in-memory wallet whose purpose is to provide basic test wallet functionality to the harness.

- Added package: `integration/harness/testnode/`
Offers a default test node that launches a new `dcrd`-instance using command-line call.

- Added package: `integration/harness/simpleregtest/`
Harbors original tests from the `rpctest`-package and offers test setup to run these tests.

### Tests:
image

### CODE REVIEW

**Comment by JFixby** [comment]
@davecgh All tests passed. You can start browsing the code.

File: integration/helpers.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+			err := os.Mkdir(pathI, 0755)
+			CheckTestSetupMalfunction(err)
+		}
+	}
+}
+
+// DeleteFile ensures file was deleted
+// reports test setup malfunction otherwise
+func DeleteFile(file string) {
+	fmt.Println("delete: " + file)
```
Comment (dajohi) [inline integration/helpers.go]: Remove fmt.Println

File: integration/harness/simpleregtest/joinnodes_test.go [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+	r := ObtainHarness(mainHarnessName)
+
+	// Create a second harness with only the genesis block so it is behind
+	// the main harness.
+	h := testSetup.Regnet0.NewInstance("checkJoinBlocks").(*harness.Harness)
+	defer testSetup.Regnet0.Dispose(h)
+
+	nodeSlice := []*harness.Harness{r, h}
+	blocksSynced := make(chan struct{})
+	go func() {
```
Comment (dajohi) [inline integration/harness/simpleregtest/joinnodes_test.go]:
```
harness/simpleregtest/joinnodes_test.go:32:2: the goroutine calls T.Fatalf, which must be called in the same goroutine as the test (staticcheck)
        go func() {
        ^
harness/simpleregtest/joinnodes_test.go:123:2: the goroutine calls T.Fatalf, which must be called in the same goroutine as the test (staticcheck)
        go func() {
        ^
harness/simpleregtest/joinnodes_test.go:145:2: the goroutine calls T.Fatalf, which must be called in the same goroutine as the test (staticcheck)
        go func() {
        ^
```

File: integration/harness/memwallet/memwallet.go [hunk 3]
```diff
@@ -205,59 +195,86 @@ func (m *memWallet) IngestBlock(header []byte, filteredTxns [][]byte) {
 	// available. We do this in a new goroutine in order to avoid blocking
 	// the main loop of the rpc client.
 	go func() {
-		m.chainUpdateSignal <- struct{}{}
+		m.chainUpdateSignal <- chainUpdateSignal
 	}()
 }
 
+// ingestBlock updates the wallet's internal utxo state based on the outputs
```
Comment (dajohi) [inline integration/harness/memwallet/memwallet.go]:
```
harness/memwallet/memwallet.go:204:31: func `(*InMemoryWallet).ingestBlock` is unused (unused)
func (wallet *InMemoryWallet) ingestBlock(update *chainUpdate) {
                              ^
harness/memwallet/memwallet.go:346:31: func `(*InMemoryWallet).unwindBlock` is unused (unused)
func (wallet *InMemoryWallet) unwindBlock(update *chainUpdate) {
                              ^
```

**Comment by JFixby** [comment]
I;ll look into it later.

**Comment by JFixby** [comment]
I did re-base. It seems to be working.

@dajohi I need some guidance on how to reproduce your comments above.

**Comment by JFixby** [comment]
FYI: The next version of this PR is released here (as a separated repo): https://github.com/JFixby/dcrregtest

**Comment by davecgh** [comment]
This is no longer relevant since `rpctest` has been split into its own [dcrtest](https://github.com/decred/dcrtest) repo.

**Pull request closed without merging**